File: day2.py

reports = []
cnt = 0
with open('day2-in.txt', 'r') as file:
    for line in file:
        report = [int(x) for x in line.strip().split()]
        reports.append(report)

for report in reports:
    safe = True
    is_asc = None
    for i in range(1, len(report)):
        diff = report[i] - report[i-1]
        if diff < -3 or diff == 0 or diff > 3:
            safe = False
            break
        if is_asc is None:
            if diff > 0:
                is_asc = True
            else:
                is_asc = False
        elif (is_asc and diff < 0) or (not is_asc and diff > 0):
            safe = False
            break
    if safe:
        cnt += 1

print(cnt)

File: day2.cpp
#include <iostream>
#include <sstream>
#include <string>

using namespace std;

bool is_safe(vector<int> &v)
{
  int len = v.size();
  bool is_asc;
  for (int i = 1; i < len; i++)
  {
    int diff = v[i] - v[i - 1];
    if (diff < -3 || diff == 0 || diff > 3)
    {
      return false;
    }
    if (i == 1)
    {
      is_asc = diff > 0;
    }
    else
    {
      if (is_asc && diff < 0 || !is_asc && diff > 0)
      {
        return false;
      }
    }
  }
  return true;
}

int main()
{
  string line;
  int cnt = 0;
  while (getline(cin, line))
  {
    stringstream ss(line);
    string s;
    vector<int> v;
    while (getline(ss, s, ' '))
    {
      int num = stoi(s);
      v.push_back(num);
    }
    cnt += is_safe(v) ? 1 : 0;
  }
  cout << cnt << endl;
  return 0;
}